src/Symfony/Component/DependencyInjection/ServiceLocator.php

```diff
@@ -45,23 +48,98 @@ public function has($id)
     public function get($id)
     {
         if (!isset($this->factories[$id])) {
-            throw new ServiceNotFoundException($id, null, null, array_keys($this->factories));
+            throw new ServiceNotFoundException($id, end($this->loading) ?: null, null, array(), $this->createServiceNotFoundMessage($id));
         }
 
-        if (true === $factory = $this->factories[$id]) {
-            throw new ServiceCircularReferenceException($id, array($id, $id));
+        if (isset($this->loading[$id])) {
+            $ids = array_values($this->loading);
+            $ids = array_slice($this->loading, array_search($id, $ids));
+            $ids[] = $id;
+
+            throw new ServiceCircularReferenceException($id, $ids);
         }
 
-        $this->factories[$id] = true;
+        $this->loading[$id] = $id;
         try {
-            return $factory();
+            return $this->factories[$id]();
         } finally {
-            $this->factories[$id] = $factory;
+            unset($this->loading[$id]);
         }
     }
 
     public function __invoke($id)
     {
         return isset($this->factories[$id]) ? $this->get($id) : null;
     }
+
+    /**
+     * @internal
+     */
+    public function withContext($externalId, Container $container)
+    {
+        $locator = clone $this;
+        $locator->externalId = $externalId;
+        $locator->container = $container;
+
+        return $locator;
+    }
+
+    private function createServiceNotFoundMessage($id)
+    {
+        if ($this->loading) {
+            return sprintf('The service "%s" has a dependency on a non-existent service "%s". This locator %s', end($this->loading), $id, $this->formatAlternatives());
+        }
+       
+        $class = debug_backtrace(DEBUG_BACKTRACE_PROVIDE_OBJECT, 3);
```

Comment:
with `DEBUG_BACKTRACE_IGNORE_ARGS` or is it negligible?

Reply:
good idea, added